### ISSUE TITLE
API proposal: Stream with BaseStream

### ISSUE THREAD

**Issue by zivkan** [dotnet/runtime#31002]
There are multiple classes that extend `Stream` and add a `BaseStream` property, such as [`DeflateStream`](https://github.com/dotnet/corefx/blob/e667c29636a622eb4f9493f75232b44e0ae90b29/src/System.IO.Compression/src/System/IO/Compression/DeflateZLib/DeflateStream.cs#L120), [`GzipStream`](https://github.com/dotnet/corefx/blob/e667c29636a622eb4f9493f75232b44e0ae90b29/src/System.IO.Compression/src/System/IO/Compression/GZipStream.cs#L170) and [`BrotliStream`](https://github.com/dotnet/corefx/blob/e667c29636a622eb4f9493f75232b44e0ae90b29/src/System.IO.Compression.Brotli/src/System/IO/Compression/BrotliStream.cs#L131).

I think this would be useful for applications that want to monitor network performance, such as download managers (I work on NuGet). Coupled with dotnet/runtime#31000 it would become very easy to measure the number of bytes downloaded, taking into account compressed responses.

## Rationale and Usage
Currently to find the base stream, my method would need to either cast to each type known in advance that has a base stream, or use reflection to look for a property named `BaseStream`. If this proposal is implemented, then it would be possible to simply cast to a single type and get the property directly.

An example of this would be getting the base stream from an HttpClient response:

```csharp
var response = await client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead);
var stream = await response.Content.ReadAsStreamAsync();

Stream baseStream;
if (stream is GZipStream gzipStream)
{
    baseStream = gzipStream.BaseStream;
}
else if (stream is DeflateStream deflateStream)
{
    baseStream = deflateStream.BaseStream;
}
else if (stream is BrotliStream brotliStream)
{
    baseStream = brotliStream.BaseStream;
}
```

With this proposal, it would be simplified to this:
```csharp
var response = await client.GetAsync(uri, HttpCompletionOption.ResponseHeadersRead);
var stream = await response.Content.ReadAsStreamAsync();

Stream baseStream;
if (stream is IStreamWrapper streamWrapper)
{
    baseStream = streamWrapper.BaseStream;
}
```

This is not only less code, but it's also adaptable to `HttpClient` implementing new compression types.

## Proposed API
```csharp
public interface IStreamWrapper
{
    Stream BaseStream { get; }
}
```

`DeflateStream`, `GzipStream` and `BrotliStream` should be changed to extend this class, but there may be others that are relevant as well.

**Comment by JeremyKuhne** [comment]
Adding a new base class wouldn't always be implementable, an interface would:

``` C#
public interface IStreamWrapper
{
   public Stream BaseStream { get; }
}

// So ...
public class GZipStream : Stream, IStreamWrapper {}
public class MySuperExtendedStreamWrapper : MyExtendedStreamWrapper, IStreamWrapper {}

// Potentially?
public class StreamReader : TextReader, IStreamWrapper {}

// etc ...

```

@zivkan Would that meet your needs?

**Comment by zivkan** [comment]
yes, that would still make it easy to find the base stream for the 3 HTTP compressed stream types. Proposing an abstract class was quite an oversight on my part. Oops.

**Comment by JeremyKuhne** [comment]
@zivkan Can you update the proposal to an interface then?

**Comment by zivkan** [comment]
Done.
